### PR TITLE
produce previousPaths metadata

### PR DESCRIPTION
**Pull request by ryanfkeepers**
## Description

Adds an additional metadata collection: a folder
id to path string mapping.  This collection is
created on backup, and retrieved along with
the delta metadata on the next backup, but is
not yet parsed or utilzed downstream.

## Type of change

- [x] :sunflower: Feature

## Issue(s)

* #1726

## Test Plan

- [x] :zap: Unit test
- [x] :green_heart: E2E

### CODE REVIEW

File: src/internal/connector/exchange/service_iterators.go [hunk 1]
```diff
@@ -81,84 +32,110 @@ func FilterContainersAndFillCollections(
...
 		edc.jobs = append(edc.jobs, jobs...)
 
 		if len(delta) > 0 {
-			deltaURLs[dirID] = delta
+			deltaURLs[cID] = delta
+		}
+
+		// add the current path for the container ID to be used in the next backup
+		// as the "previous path", for reference in case of a rename or relocation.
+		if itp, err := resolver.IDToPath(ctx, cID); err != nil {
```
Comment (ashmrtn) [inline src/internal/connector/exchange/service_iterators.go:89]: I think `dirPath` from above already has what this is doing?

File: src/internal/connector/exchange/service_iterators.go [hunk 2]
```diff
@@ -81,84 +32,110 @@ func FilterContainersAndFillCollections(
...
+			errs = support.WrapAndAppend(qp.ResourceOwner, err, errs)
+		} else {
+			// Some root folders have no display name. The exchange path will
+			// fail without that, so we inject a fallback here just for
+			// the previousPaths handling.
+			els := itp.Elements()
+			if len(els) == 0 || len(els[len(els)-1]) == 0 {
+				itp = itp.Append(rootFolderAlias)
```
Comment (ashmrtn) [inline src/internal/connector/exchange/service_iterators.go:97]: _Probably_ ok to skip this? I don't think there's any cases where we persist the root directory, the root directory has items, and we don't give the root directory a path

Comment (ashmrtn) [inline src/internal/connector/exchange/service_iterators.go:97]: `rootFolderAlias` refers specifically to the Exchange mail root folder. Should probably be renamed to reflect that in a different PR eventually

File: src/internal/connector/exchange/service_iterators.go [hunk 3]
```diff
@@ -81,84 +32,110 @@ func FilterContainersAndFillCollections(
...
+			p, err := itp.ToDataLayerExchangePathForCategory(
+				qp.Credentials.AzureTenantID,
+				qp.ResourceOwner,
+				qp.Category,
+				false)
+			if err != nil {
+				errs = support.WrapAndAppend(qp.ResourceOwner, err, errs)
+			} else {
+				prevPaths[cID] = p.Folder()
+			}
```
Comment (ashmrtn) [inline src/internal/connector/exchange/service_iterators.go:109]: this can also be removed because we can use `dirPath` (I think)
```suggestion
```

File: src/internal/connector/exchange/service_iterators.go [hunk 4]
```diff
@@ -81,84 +32,110 @@ func FilterContainersAndFillCollections(
...
+		graph.PreviousPathFileName,
+		path.ExchangeService,
+		qp.Category,
+		prevPaths,
+		statusUpdater,
+	); err != nil {
+		errs = support.WrapAndAppend("making paths metadata collection", err, errs)
 	} else if col != nil {
-		collections[metadataKey] = col
+		collections[graph.PathsMetadataCollectionKey] = col
```
Comment (ashmrtn) [inline src/internal/connector/exchange/service_iterators.go:138]: this is going to cause undefined behavior in KW when it tries to consume it. We now have two collections that have the same `FullPath()`. We should instead be creating a single metadata collection with two files in it

File: src/internal/connector/graph/service.go [hunk 5]
```diff
@@ -9,9 +9,23 @@ import (
...
+	PreviousPathFileName = "previouspath"
+)
+
+const (
+	// DeltaMetadataCollectionKey is the name of the collection holding the delta
+	// url lookups inside a collection map otherwise keyed by folder ids.
+	DeltaMetadataCollectionKey = "delta-metadata"
+	// PathsMetadataCollectionKey is the name of the collection holding the previous
+	// paths lookups inside a collection map otherwise keyed by folder ids.
+	PathsMetadataCollectionKey = "paths-metadata"
```
Comment (ashmrtn) [inline src/internal/connector/graph/service.go:27]: this is just an artifact of needing to have some map key so we can add it to the set of collections the exchange code deals with. It could really be a single map key (since we should only be making one collection) and it should be in the exchange files not here (it's not something that will be needed outside of backup). Though, it may also be possible to use the `col.FullPath().ShortRef()` or `col.FullPath().String()` as the map key for the metadata collection

The map was used just to make sure that we only had one instance of a collection for each path when we were fetching items in a different way. It's probably not really required anymore but that code hasn't been refactored in a while

File: src/internal/connector/graph/service.go [hunk 6]
```diff
@@ -9,9 +9,23 @@ import (
...
-// DeltaTokenFileName is the name of the file containing delta token(s) for a
-// given endpoint. The endpoint granularity varies by service.
-const DeltaTokenFileName = "delta"
+const (
+	// DeltaTokenFileName is the name of the file containing delta token(s) for a
+	// given endpoint. The endpoint granularity varies by service.
+	DeltaTokenFileName = "delta"
+	// PreviousPathFileName is the name of the file containing previous path(s) for a
+	// given endpoint.
+	PreviousPathFileName = "previouspath"
```
Comment (ashmrtn) [inline src/internal/connector/graph/service.go:18]: This should be added to the return value of `MetadataFileNames` eventually

File: src/internal/connector/exchange/service_iterators.go (hunk 1 above)
Comment (ryanfkeepers) [reply src/internal/connector/exchange/service_iterators.go:89]: Oh... yes.  Yes it does.

File: src/internal/connector/exchange/service_iterators.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/connector/exchange/service_iterators.go:97]: Fwiw, it wasn't about consistency so much as getting yelled at by the builder for having zero elements.  But it's all good, because I can just use the `dirPath`!

File: src/internal/connector/exchange/service_iterators.go [hunk 7]
```diff
@@ -81,84 +32,110 @@ func FilterContainersAndFillCollections(
...
+
+			p, err := itp.ToDataLayerExchangePathForCategory(
+				qp.Credentials.AzureTenantID,
+				qp.ResourceOwner,
+				qp.Category,
+				false)
+			if err != nil {
+				errs = support.WrapAndAppend(qp.ResourceOwner, err, errs)
+			} else {
+				prevPaths[cID] = p.Folder()
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/connector/exchange/service_iterators.go:108]: I was thinking about this, and that we'd probably be safer overall storing slice over the folder as a string.  Saves us from parsing collisions like this.  Something I'll change in a follow-up, to not overwhelm this PR.

File: src/internal/connector/graph/service.go (hunk 6 above)
Comment (ryanfkeepers) [reply src/internal/connector/graph/service.go:18]: whoops, good catch!

File: src/internal/connector/exchange/service_iterators.go (hunk 4 above)
Comment (ryanfkeepers) [reply src/internal/connector/exchange/service_iterators.go:138]: Bother.  I knew something about that was tickling my brain.  Should be fixed now.

File: src/internal/connector/exchange/service_iterators.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ashmrtn) [reply src/internal/connector/exchange/service_iterators.go:97]: mmm so there's a subtle difference here. If you just use `path.Builder` (what the `CachedContainer.Path()` returns then it can be empty. If you use a `path.Path` implementation then it will have the prefix elements (tenant, service, etc) and require at least one folder or item name element. The `ToDataLayerX()` functions create `path.Path` instances

File: src/internal/connector/exchange/service_iterators.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ashmrtn) [reply src/internal/connector/exchange/service_iterators.go:108]: >safer overall storing slice over the folder as a string.

I'm not sure I understand this. Can you explain more or point me to the code when you have it if it's going to be in this PR

[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/connector/exchange/service_iterators.go:108]: Ah, sorry, late night brain was getting to me.  I did half a read and then misinterpreted your comment as worrying about string->path parsing.  The test is a good idea, and I'll add it in, but personally I have very few worries about json decoding/encoding. Or, at least, the golang package for it.  If graph has its own bugs there I'm not aware of them.

[the remark this answers is not in the document]
Comment (ashmrtn) [reply src/internal/connector/exchange/service_iterators.go:108]: mmm I don't think it would be a bug in the golang JSON parser. I'm more worried about us not properly escaping things we need to when handing data to it and that causing problems down the line. It's more of a misunderstanding of the API and assumptions than a bug

File: src/internal/connector/graph/metadata_collection.go [hunk 8]
```diff
@@ -88,7 +163,7 @@ func (md MetadataCollection) Items() <-chan data.Stream {
 	return res
 }
 
-// MetadataItem is an in-memory data.Stream implementation. MetadataItem does
+// metadataItem is an in-memory data.Stream implementation. MetadataItem does
```
Comment (ashmrtn) [inline src/internal/connector/graph/metadata_collection.go:166]: Not sure if this was meant to be changed to unexported or not
```suggestion
// MetadataItem is an in-memory data.Stream implementation. MetadataItem does
```

File: src/internal/connector/graph/service.go [hunk 9]
```diff
@@ -9,14 +9,30 @@ import (
...
 // metadata such as delta tokens and folderID->path references.
-func MetadataFileNames() []string {
-	return []string{DeltaTokenFileName}
+func AllMetadataFileNames() []string {
+	return []string{DeltaTokenFileName, PreviousPathFileName}
+}
+
+// MetadataFileNames produces the category-specific set of filenames used to
+// store graph metadata such as delta tokens and folderID->path references.
+func MetadataFileNames(cat path.CategoryType) []string {
```
Comment (ashmrtn) [inline src/internal/connector/graph/service.go:29]: this feels like eventually it should live in the exchange package or something, but for now this works

**Review by ashmrtn — APPROVED** [review]

File: src/internal/connector/graph/service.go (hunk 9 above)
Comment (ryanfkeepers) [reply src/internal/connector/graph/service.go:29]: Sure, why not.

File: src/internal/connector/graph/metadata_collection.go (hunk 8 above)
Comment (ryanfkeepers) [reply src/internal/connector/graph/metadata_collection.go:166]: Oh yeah, i went back and forth on that a few times.

File: src/internal/connector/exchange/service_iterators.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/connector/exchange/service_iterators.go]: Topic resolved in sync discussion; appropriate tests added.

File: src/internal/connector/exchange/service_iterators.go [hunk 10]
```diff
@@ -83,85 +34,83 @@ func FilterContainersAndFillCollections(
...
-
-		jobs, delta, err := fetchFunc(ctx, edc.service, qp.ResourceOwner, dirID, oldDelta)
+		jobs, delta, err := fetchFunc(ctx, edc.service, qp.ResourceOwner, cID, oldDeltas[cID])
 		if err != nil {
-			errs = support.WrapAndAppend(
-				qp.ResourceOwner,
-				err,
-				errs,
-			)
+			errs = support.WrapAndAppend(qp.ResourceOwner, err, errs)
```
Comment (dadams39) [inline src/internal/connector/exchange/service_iterators.go:81]: Should a continue be added following this error?
